### PR TITLE
Basic unit tests, logging, exception refactoring

### PR DESCRIPTION
**Pull request by cfculhane**
This PR combines a bunch of commits to address the following issues:

#8 - Implemented via shared_logging.py and its associated config file, at the moment the level is set to DEBUG which is quite verbose, but once the codebase becomes more stable this can be rolled back if desired

#22 - Using pytest and tox, I have written a basic integration test that runs two short videos and checks they are outputted correctly and that the output contains the expected number of frames. Later on further test can be written to test individual modules but at least now we can easily check if we break things as we change the code.

I have included calls to plugins to make coverage and test reports, see below:

image

image

#23 - partial fix to this, but the whole stack of .arm(), .release() etc might need to be looked at in more detail to figure out how untangle it and ensure things exit cleanly. For now I've removed the sys.exit() calls and made the exit clean by calling .release() when running out of frames. This might not be the intended behaviours, so will need feedback.

Please have a look through and see how I went, the blink_detection is a  bit of an issue I think, I've added an assert that the blink_threshold > 0, as there is a check for abs(delta) > threshold if the threshold is less than 0, check_blink() will always detect a blink. #24 raises the issue of not hardcoding this, which might need to be addressed in a PR soon after this one.

### CODE REVIEW

**Comment by cfculhane** [comment]
tox doesn't want to start on travis for some reason! :(

File: setup.py [hunk 1]
```diff
@@ -1,8 +1,7 @@
 #!/usr/bin/env python
-from setuptools import setup, find_packages
+from setuptools import setup, find_namespace_packages
```
Comment (kinow) [inline setup.py]: I think it needs to be `find_packages`? Unless we are now using namespace packages (e.g. `google.subpackageA`, `google.subpackageB`, etc, PEP-420).

**Review by kinow — COMMENTED** [review]
After changing the `first_namespace_packages` by `find_packages` it executed `tox`, then failed with the same message as Travis.

```
------------------------------------------------------------------------------- live log call --------------------------------------------------------------------------------
2020-07-19 20:53:22 INFO Initiating tracking via Importer: cv
ERROR: InvocationError for command /home/kinow/Development/python/workspace/eyeloop/.tox/py38/bin/pytest --cov=eyeloop --cov-report html:tests/reports/coverage --html=tests/reports/pytest_results.html (exited with code 1)
___________________________________________________________________________________ summary ___________________________________________________________________________________
ERROR:   py38: commands failed

```

Let me see if there's any easy flag to increase verbosity...

**Comment by kinow** [comment]
@cfculhane I think it needs an X Server to run the tests? I ran `tox` and it failed. Then copied the command, it opened lots of windows, but the tests started working (I think; at least it went a bit ahead until I stopped it as it was really opening a lot of windows :smile: )

File: setup.py (hunk 1 above)
Comment (cfculhane) [reply setup.py]: Hmm, I changed this as it allowed `tox` to more easily find the submodules, it was after an hour of frustration. I've reverted this change and it works, so maybe I changed something else that fixed it! Good pickup :)

**Comment by cfculhane** [comment]
> @cfculhane I think it needs an X Server to run the tests? I ran `tox` and it failed. Then copied the command, it opened lots of windows, but the tests started working (I think; at least it went a bit ahead until I stopped it as it was really opening a lot of windows 😄 )

I've found, https://github.com/tox-dev/tox-travis , ill add that to the travis script to see what happens

**Comment by cfculhane** [comment]
Still not working, I'll try this approach - https://github.com/jazzband/pip-tools/issues/858

**Comment by cfculhane** [comment]
Still no dice, open to ideas. Maybe it does need an X server and we could just run travis without a GUI - but given it was running eyeloop previously this doesn't seem to be the cause

**Comment by kinow** [comment]
I'm not in front of the computer with my development environment, but I found this link from pytest-qt when searching for xfvb and tox. Check out the first error described, looks pretty similar to what we have here.

https://pytest-qt.readthedocs.io/en/latest/troubleshooting.html

Perhaps their solution could be ported to eyeloop? Or at least serve as starting point I think.

**Comment by cfculhane** [comment]
> I'm not in front of the computer with my development environment, but I found this link from pytest-qt when searching for xfvb and tox. Check out the first error described, looks pretty similar to what we have here.
> 
> [pytest-qt.readthedocs.io/en/latest/troubleshooting.html](https://pytest-qt.readthedocs.io/en/latest/troubleshooting.html)
> 
> Perhaps their solution could be ported to eyeloop? Or at least serve as starting point I think.

Thanks for this, I'll look into it later in the week, I've unfortunately run out of time today

**Comment by cfculhane** [comment]
Hallelujah, just using pytest works! Tox is still nice for easy of use by dev and end users I think, but running pytest seems to be the go for travis for now.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# tox (https://tox.readthedocs.io/) is a tool for running tests
```
Comment (kinow) [inline tox.ini:1]: From your last comment it looks like Travis will be running `pytest` now @cfculhane? Do we still need `tox.ini`?

File: setup.py [hunk 3]
```diff
@@ -1,8 +1,7 @@
 #!/usr/bin/env python
 from setuptools import setup, find_packages
 
-
-install_requires = install_requires = []
```
Comment (kinow) [inline setup.py:5]: Oh! Good catch :+1:

File: eyeloop/utilities/shared_logging.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    log_filename = None
+
+    # Check for permissions and change log dir if write access isn't granted
+    try:
+        if Path(log_dir).exists() is False:
+            print(f"log dir not found, Attempting to create dir {log_dir}")
+            Path(log_dir).mkdir(parents=True, exist_ok=True)
+    except PermissionError:
+        log_dir = Path("C:/Temp", "coronaSEIR", "logs")
+        print(f"WARNING: Not enough permissions to write to log_dir given. Writing to {log_dir} instead.")
```
Comment (kinow) [inline eyeloop/utilities/shared_logging.py]: Will this work on Linux? That `C:/Temp` I mean?

**Review by kinow — APPROVED** [review]
That's looking good to me. A great improvement! And users (and devs) should be able to change the logging configuration and see more - or less - of the eyeloop logging information. :clap:

File: tox.ini (hunk 2 above)
Comment (cfculhane) [reply tox.ini:1]: Yeah I think tox is easier for end users, up to you though. I like to use tox to test fresh installs, and it's easy to test across multiple environments as it creates its own tox venv's for each python version.

File: eyeloop/utilities/shared_logging.py (hunk 4 above)
Comment (cfculhane) [reply eyeloop/utilities/shared_logging.py]: Oops this is what happens when you copy code from your previous project ;) I'll remove this now, permission error's shouldn't occur writing logs to the /data/ directory as intended in this application, so I'll remove the try: except.

**Comment by simonarvin** [comment]
Great work, guys! I will look into blink detection. Closing Importer via .release() is more elegant - I fully agree. Merging this into our main branch

**Pull request merged**